pkg/trace/api/evp_intake_proxy.go

```diff
@@ -0,0 +1,234 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		msg := fmt.Sprintf("EVPProxy is disabled: %v", message)
+		http.Error(w, msg, http.StatusMethodNotAllowed)
+	})
+}
+
+// evpProxyForwarder creates an http.ReverseProxy which can forward payloads to
+// one or more endpoints, based on the request received and the Agent configuration.
+// Headers are not proxied, instead we add our own known set of headers.
+// See also evpProxyTransport below.
+func evpProxyForwarder(conf *config.AgentConfig, endpoints []config.Endpoint, transport http.RoundTripper, logger *stdlog.Logger) http.Handler {
+	return &httputil.ReverseProxy{
+		Director: func(req *http.Request) {
+			req.Header["X-Forwarded-For"] = nil // Prevent setting X-Forwarded-For
+		},
+		ErrorLog:  logger,
+		Transport: &evpProxyTransport{transport, endpoints, conf},
+	}
+}
+
+// evpProxyTransport sends HTTPS requests to multiple targets using an
+// underlying http.RoundTripper. API keys are set separately for each target.
+// When multiple endpoints are in use the response from the first endpoint
+// is proxied back to the client, while for all aditional endpoints the
+// response is discarded.
+type evpProxyTransport struct {
+	transport http.RoundTripper
+	endpoints []config.Endpoint
+	conf      *config.AgentConfig
+}
+
+func (t *evpProxyTransport) RoundTrip(req *http.Request) (rresp *http.Response, rerr error) {
+	if req.Body != nil && t.conf.EVPProxy.MaxPayloadSize > 0 {
+		req.Body = apiutil.NewLimitedReader(req.Body, t.conf.EVPProxy.MaxPayloadSize)
+	}
+
+	// Metrics with stats for debugging
+	beginTime := time.Now()
+	metricTags := []string{}
+	if ct := req.Header.Get("Content-Type"); ct != "" {
+		metricTags = append(metricTags, "content_type:"+ct)
+	}
+	defer func() {
+		metrics.Count("datadog.trace_agent.evp_proxy.request", 1, metricTags, 1)
+		metrics.Count("datadog.trace_agent.evp_proxy.request_bytes", req.ContentLength, metricTags, 1)
+		metrics.Timing("datadog.trace_agent.evp_proxy.request_duration_ms", time.Since(beginTime), metricTags, 1)
+		if rerr != nil {
+			metrics.Count("datadog.trace_agent.evp_proxy.request_error", 1, metricTags, 1)
+		}
+	}()
+
+	// Read the input headers
+	subdomain := req.Header.Get("X-Datadog-EVP-Subdomain")
+	containerID := req.Header.Get("Datadog-Container-ID")
+	contentType := req.Header.Get("Content-Type")
+	userAgent := req.Header.Get("User-Agent")
+
+	// Sanitize the input
+	if len(subdomain) == 0 {
+		return nil, fmt.Errorf("EVPProxy: no subdomain specified")
+	}
+	if !isValidSubdomain(subdomain) {
+		return nil, fmt.Errorf("EVPProxy: invalid subdomain: %s", subdomain)
+	}
+	metricTags = append(metricTags, "subdomain:"+subdomain)
+	if !isValidPath(req.URL.Path) {
+		return nil, fmt.Errorf("EVPProxy: invalid target path: %s", req.URL.Path)
+	}
+	if !isValidQueryString(req.URL.RawQuery) {
+		return nil, fmt.Errorf("EVPProxy: invalid query string: %s", req.URL.RawQuery)
+	}
```

Comment:
We don't need to support all possible paths and domains, so we are limiting this further than what Go does. We are not accepting any valid path.

Reply:
In that case, can you please document the helper functions so that it is clear how they are different from regular checks and why they are needed? Also, please put the helper functions at the end of the file, so we can start with the main structures and logic.